src/classes/PMT_Payment_TDTM.cls

```diff
@@ -201,4 +162,132 @@ public class PMT_Payment_TDTM extends TDTM_Runnable {
+        DmlWrapper dmlWrapper) {
+
+        if (dmlWrapper == null) {
+            dmlWrapper = new DmlWrapper();
+        }
+
+        Set<Id> oppIds = new Set<Id>();
+        for (npe01__OppPayment__c pmt : newlist) {
+            oppIds.add(pmt.npe01__Opportunity__c);
+        }
+
+        Map<Id, Opportunity> mapOfOpps = getPaymentOpportunities(oppIds);
+        
+        // Validation and default the CurrencyIsoCode for new records for Before Insert/update
+        if (isCurrencyFieldAvailable && (triggerAction == TDTM_Runnable.Action.BeforeInsert || triggerAction == TDTM_Runnable.Action.BeforeUpdate)) {
+            for (Integer i = 0; i < newlist.size(); i++) {
+                npe01__OppPayment__c pmt = newlist[i];
+                String oldPmtCurrency = (oldList != null ? (String) oldlist[i].get('CurrencyIsoCode') : null);
+                setPaymentCurrency(mapOfOpps, pmt, oldPmtCurrency, triggerAction);
+            }
+        }
+
+        // For an After Insert/Update Payment action, when Advanced Curr Mgt is enabled and there is a defined
+        // Closed stage name in settings, roll up the total paynents and compare to the Opp.Amount. If the
+        // two values are equal then close the Opportunity by setting the stage name.
+        if (triggerAction == TDTM_Runnable.Action.AfterInsert || triggerAction == TDTM_Runnable.Action.AfterUpdate) {
+
+            npe01__Contacts_And_Orgs_Settings__c ContactSettings = UTIL_CustomSettingsFacade.getContactsSettings();
+            String closedStageName = ContactSettings.Payments_Auto_Close_Stage_Name__c;
+
+            Boolean isACMEnabled = UTIL_Currency.getInstance().isAdvancedCurrencyManagementEnabled();
+            if (isACMEnabled && String.isNotEmpty(closedStageName)) {
+                // Loop through all the queried Opps with the subquery on payments
+                for (Opportunity opp : mapOfOpps.values()) {
+                   handleAdvancedCurrencyPayments(opp, closedStageName, dmlWrapper);
+                }
+            }
+        }
+        return dmlWrapper;
+    }
+
+    /*******************************************************************************************************
+    * @description              For multicurrency orgs, set the currency of the payment to the currency of 
+    *                           the opportunity and make sure that the payment currency was not modified
+    * @param mapOfOpps          Map of Opportunities
+    * @param pmt                Payment Object
+    * @param oldPmtCurrency     String of old Payment Currency
+    * @param triggerAction      Action that triggered this execution
+    * @return void              Does not return anything
+    ********************************************************************************************************/
+    private void setPaymentCurrency(Map<Id, Opportunity> mapOfOpps, npe01__OppPayment__c pmt, String oldPmtCurrency, TDTM_Runnable.Action triggerAction) {
```

Comment:
What do you think about `validateAndSetPaymentCurrency()` as the method name?

Reply:
renamed